File: apps/android/data/src/main/kotlin/com/sanvya/app/data/repository/InvestmentsRepository.kt
package com.sanvya.app.data.repository

/**
 * Investments (holdings) repository -- P3.20 (task #26), read/write.
 * Mirrors apps/web/src/investments/write.ts's addHolding() and page.tsx's
 * EditHolding.save()/remove() exactly. See docs/mobile/screen-specs/
 * investments.md for the full source-verification notes.
 *
 * Fixed this pass (2026-08-06): `exchange`/`instrument_type`/`avg_cost` were
 * previously mapped as non-nullable via non-optional cursor getters despite
 * web's HoldingRow treating all three as nullable (`exchange: string | null`,
 * `instrument_type: string | null`, `avg_cost: number | null`) -- a real
 * runtime-crash risk for any holding where those columns are actually null
 * (e.g. a freshly-added FD/crypto/other-scheme holding has no exchange).
 * `current_value` was already correctly typed `Long?` in Kotlin but was
 * still read via non-optional `getLong`, which would crash the exact same
 * way. Also added the three columns the schema already had
 * (PocketCareSchema.kt's `holdings` TableDef) but this repository's
 * `Holding` data class was missing entirely: `off_list` (functionally
 * significant -- gates live pricing/"untracked" display, see
 * domain/investments/InvestmentsModel.kt's `holdingLabel`/`valuation`),
 * `source_account_id`, `planned_id`.
 */

import com.powersync.PowerSyncDatabase
import com.powersync.db.SqlCursor
import com.powersync.db.getBooleanOptional
import com.powersync.db.getDouble
import com.powersync.db.getDoubleOptional
import com.powersync.db.getLong
import com.powersync.db.getLongOptional
import com.powersync.db.getString
import com.powersync.db.getStringOptional
import kotlinx.coroutines.flow.Flow

/** One `market_dividends` row (global, read-only market-sync table). */
data class DividendRow(val symbol: String, val exchange: String?, val exDate: String, val payDate: String?, val amount: Long, val currency: String)

/** One `market_quotes` row (global, read-only). */
data class QuoteRow(val symbol: String, val exchange: String?, val price: Long, val currency: String)

data class Holding(
    val id: String,
    val userId: String,
    val accountId: String,
    val symbol: String,
    val exchange: String?,
    val quantity: Double,
    val avgCost: Long?,
    val currency: String,
    val autoFetch: Boolean,
    val instrumentType: String?,
    val offList: Boolean,
    val name: String?,
    val assetClass: String?,
    val currentValue: Long?,
    val annualRate: Double?,
    val maturityDate: String?,
    val sourceAccountId: String?,
    val plannedId: String?,
    /** Amount-based SIP fields (migration 0061). `plannedId` alone is not
     * enough to tell a live SIP from a stopped one -- it can still point at a
     * recurring row that was already cancelled -- so web gates the SIP UI on
     * `planned_id && sip_amount > 0` and so does mobile. */
    val sipAmount: Long?,
    val sipDay: Long?,
)

/** Funding mode for a new holding -- matches web's AddHoldingInput.funding
 * union: an EXISTING holding raises the investment account's invested pool
 * via an `adjustment` entry (money was already in the market); a NEW
 * investment `transfer`s the cost from a chosen savings/other account into
 * the investment account (money moves, net worth preserved). */
sealed class HoldingFunding {
    object Existing : HoldingFunding()
    data class New(val sourceAccountId: String) : HoldingFunding()
}

/**
 * A SIP the user is setting up alongside the holding -- web's
 * `AddHoldingInput.sip`.
 *
 * [firstDue] is the date the recurring engine posts the first instalment and
 * [startDate] is when the plan began; web sends the same value for both from
 * one date field, and they are kept separate here because the COLUMNS are
 * separate (`recurring_items.next_due` moves every time the engine posts,
 * `holdings.sip_start_date` never does).
 *
 * [day] is the day-of-month, already clamped to 1-28 by
 * `domain.investments.clampSipDay` -- the repository does not re-clamp, so
 * there is exactly one place the rule lives.
 */
data class SipSetup(
    val amount: Long,
    val frequency: String,
    val firstDue: String,
    val sourceAccountId: String,
    val startDate: String,
    val day: Int,
)

data class AddHoldingInput(
    val investmentAccountId: String,
    val assetClass: String,
    val symbol: String,
    val exchange: String?,
    val name: String,
    val quantity: Double,
    val avgCost: Long?,
    val currency: String,
    val currentValue: Long?,
    val annualRate: Double?,
    val maturityDate: String?,
    val offList: Boolean,
    val autoFetch: Boolean,
    val funding: HoldingFunding,
    /** Present only when the user is starting a SIP. See [SipSetup]. */
    val sip: SipSetup? = null,
)

class InvestmentsRepository(private val db: PowerSyncDatabase) {

    private fun holdingMapper(cursor: SqlCursor): Holding = Holding(
        id = cursor.getString("id"),
        userId = cursor.getString("user_id"),
        accountId = cursor.getString("account_id"),
        symbol = cursor.getString("symbol"),
        exchange = cursor.getStringOptional("exchange"),
        quantity = cursor.getDouble("quantity"),
        avgCost = cursor.getLongOptional("avg_cost"),
        currency = cursor.getString("currency"),
        autoFetch = cursor.getBooleanOptional("auto_fetch") ?: false,
        instrumentType = cursor.getStringOptional("instrument_type"),
        offList = cursor.getBooleanOptional("off_list") ?: false,
        name = cursor.getStringOptional("name"),
        assetClass = cursor.getStringOptional("asset_class"),
        currentValue = cursor.getLongOptional("current_value"),
        annualRate = cursor.getDoubleOptional("annual_rate"),
        maturityDate = cursor.getStringOptional("maturity_date"),
        sourceAccountId = cursor.getStringOptional("source_account_id"),
        plannedId = cursor.getStringOptional("planned_id"),
        sipAmount = cursor.getLongOptional("sip_amount"),
        sipDay = cursor.getLongOptional("sip_day"),
    )

    fun watchHoldings(userId: String): Flow<List<Holding>> = db.watch(
        sql = "SELECT * FROM holdings WHERE deleted_at IS NULL AND user_id = ? ORDER BY created_at DESC",
        parameters = listOf(userId),
        mapper = ::holdingMapper,
    )

    /** Global, read-only market-sync tables -- no user_id/deleted_at filter,
     * matches web's own unscoped queries exactly. Added 2026-08-06 for
     * Insights' dividend_income/portfolio_projection cards (task #28), the
     * first mobile reader of either table. */
    fun watchDividends(): Flow<List<DividendRow>> = db.watch(
        sql = "SELECT symbol, exchange, ex_date, pay_date, amount, currency FROM market_dividends",
        mapper = { cursor ->
            DividendRow(
                symbol = cursor.getString("symbol"), exchange = cursor.getStringOptional("exchange"),
                exDate = cursor.getString("ex_date"), payDate = cursor.getStringOptional("pay_date"),
                amount = cursor.getLong("amount"), currency = cursor.getString("currency"),
            )
        },
    )

    fun watchQuotes(): Flow<List<QuoteRow>> = db.watch(
        sql = "SELECT symbol, exchange, price, currency FROM market_quotes",
        mapper = { cursor ->
            QuoteRow(
                symbol = cursor.getString("symbol"), exchange = cursor.getStringOptional("exchange"),
                price = cursor.getLong("price"), currency = cursor.getString("currency"),
            )
        },
    )

    /**
     * Adds a holding -- matches write.ts's addHolding() exactly: funds the
     * invested pool first (transfer from a source account, or an adjustment
     * on the investment account itself for an already-owned holding), creates
     * the SIP's recurring transfer if there is one, then inserts the
     * `holdings` row.
     *
     * The SIP half was missing until now, and its absence was not cosmetic:
     * `planned_id` was hardcoded null and `sip_amount`/`sip_day` were never
     * written, so a SIP could not be created on a phone AT ALL and the Stop
     * SIP control could only ever appear for a holding created on web
     * (PARITY_AUDIT 2026-08-28, item 9).
     */
    suspend fun addHolding(userId: String, inp: AddHoldingInput): String {
        val costTotal = Math.round((inp.avgCost ?: 0L) * inp.quantity)

        if (costTotal > 0) {
            when (inp.funding) {
                is HoldingFunding.New -> insertRow(
                    db, "transactions", userId,
                    mapOf(
                        "account_id" to inp.funding.sourceAccountId,
                        "type" to "transfer",
                        "amount" to costTotal,
                        "currency" to inp.currency,
                        "to_account_id" to inp.investmentAccountId,
                        "to_amount" to costTotal,
                        "description" to "Invested in ${inp.name.ifBlank { inp.symbol.ifBlank { "investment" } }}",
                        "occurred_at" to nowIso(),
                    ),
                )
                HoldingFunding.Existing -> insertRow(
                    db, "transactions", userId,
                    mapOf(
                        "account_id" to inp.investmentAccountId,
                        "type" to "adjustment",
                        "amount" to costTotal,
                        "currency" to inp.currency,
                        "description" to "Existing investment: ${inp.name.ifBlank { inp.symbol.ifBlank { "holding" } }}",
                        "occurred_at" to nowIso(),
                    ),
                )
            }
        }

        // The SIP is a recurring `saving` transfer: source account -> this
        // investment account, auto-posting on its own schedule.
        //
        // This is the ONLY place a recurring saving item is created on either
        // platform, exactly as on web. Recurring savings are not browsable
        // under Recurring, so a SIP belongs to the holding that funds it and
        // is created and stopped there (see stopSipForHolding). It still posts
        // through the shared engine and still shows in the "Due now" strip.
        //
        // One row, not a template + rule pair: `recurring_items` carries both
        // the schedule and the transaction detail (migration 0064).
        val plannedId: String? = inp.sip?.takeIf { it.amount > 0 && it.sourceAccountId.isNotBlank() }?.let { sip ->
            insertRow(
                db, "recurring_items", userId,
                mapOf(
                    "direction" to "saving",
                    "name" to inp.name.ifBlank { inp.symbol.ifBlank { "SIP" } },
                    "amount" to sip.amount,
                    "currency" to inp.currency,
                    "frequency" to sip.frequency,
                    // Web hardcodes 1 here too -- every interval the UI offers
                    // is "every 1 <frequency>".
                    "interval_count" to 1L,
                    "next_due" to sip.firstDue,
                    "account_id" to sip.sourceAccountId,
                    "to_account_id" to inp.investmentAccountId,
                    "category_id" to null,
                    "auto_post" to 1L,
                    "active" to 1L,
                    "alert_time_utc" to null,
                    "description" to "SIP",
                    "note" to null,
                    "payment_method" to null,
                    "labels" to null,
                    "split_group_id" to null,
                    "split_mode" to null,
                    "last_generated" to null,
                    "source_table" to null,
                    "source_id" to null,
                ),
            )
        }

        return insertRow(
            db, "holdings", userId,
            mapOf(
                "account_id" to inp.investmentAccountId,
                "symbol" to inp.symbol,
                "exchange" to inp.exchange,
                "name" to inp.name.ifBlank { null },
                "quantity" to inp.quantity,
                "avg_cost" to inp.avgCost,
                "currency" to inp.currency,
                "asset_class" to inp.assetClass,
                "instrument_type" to when (inp.assetClass) { "mf" -> "mf"; "stock" -> "stock"; else -> null },
                "current_value" to inp.currentValue,
                "annual_rate" to inp.annualRate,
                "maturity_date" to inp.maturityDate,
                // A SIP's money movement IS the recurring transfer, so its
                // source account is the debit account, not a one-off funding
                // account -- web resolves it in the same order.
                "source_account_id" to when {
                    inp.sip != null -> inp.sip.sourceAccountId
                    inp.funding is HoldingFunding.New -> inp.funding.sourceAccountId
                    else -> null
                },
                "planned_id" to plannedId,
                "off_list" to if (inp.offList) 1L else 0L,
                "auto_fetch" to if (inp.autoFetch) 1L else 0L,
                // Amount-based SIP fields (migration 0061). Written together
                // with `planned_id`: the UI reads a live SIP as
                // `planned_id != null && sip_amount > 0`, and half of that
                // pair is what made every mobile-created holding look
                // SIP-less.
                "sip_amount" to plannedId?.let { inp.sip?.amount },
                "sip_start_date" to plannedId?.let { inp.sip?.startDate },
                "sip_day" to plannedId?.let { inp.sip?.day?.toLong() },
            ),
        )
    }

    /** Matches web's EditHolding.save(): quantity/avg_cost always editable;
     * current_value/annual_rate only meaningful for unpriced holdings but
     * harmless to write through for priced ones (web does the same --
     * updateRow always sends all four). */
    suspend fun updateHolding(id: String, quantity: Double, avgCost: Long?, currentValue: Long?, annualRate: Double?) {
        updateRow(
            db, "holdings", id,
            mapOf("quantity" to quantity, "avg_cost" to avgCost, "current_value" to currentValue, "annual_rate" to annualRate),
        )
    }

    /**
     * Stop the SIP attached to a holding -- web's `stopSipForHolding()`.
     *
     * The recurring transfer is a standing debit on a real account, and
     * recurring savings are not browsable under Recurring, so if this is
     * skipped there is nowhere left for the user to go and cancel it: the
     * money keeps leaving every month for an investment that no longer
     * exists. No-op when the holding never had one.
     */
    suspend fun stopSipForHolding(plannedId: String?) {
        if (plannedId != null) softDelete(db, "recurring_items", plannedId)
    }

    /** Clears the holding's own SIP fields, so `sip_amount > 0` stops
     * reporting a live SIP once the recurring row is gone. Matches web's
     * `updateRow("holdings", id, { sip_amount: null, sip_day: null })`. */
    suspend fun clearSipFields(id: String) {
        updateRow(db, "holdings", id, mapOf<String, Any?>("sip_amount" to null, "sip_day" to null))
    }

    /**
     * Matches web's remove(): soft-delete the holding row, and kill its SIP
     * first -- the funding transaction is deliberately NOT reversed (same
     * accepted asymmetry as Goals' allocation-delete-doesn't-cascade,
     * documented in docs/mobile/screen-specs/goals.md), but a live standing
     * debit is not an asymmetry, it is money still moving.
     */
    suspend fun deleteHolding(id: String, plannedId: String?) {
        stopSipForHolding(plannedId)
        softDelete(db, "holdings", id)
    }
}

File: apps/ios/Data/Sources/Data/InvestmentsRepository.swift
import Foundation
import PowerSync

/// Investments (holdings) repository -- P3.20 (task #26), read/write.
/// Mirrors apps/web/src/investments/write.ts's addHolding() and
/// page.tsx's EditHolding.save()/remove() exactly, and Android's
/// InvestmentsRepository.kt field-for-field. See docs/mobile/
/// screen-specs/investments.md for the full source-verification notes.
///
/// Fixed this pass (2026-08-06): `exchange`/`instrumentType`/`avgCost`
/// were previously mapped as non-nullable via non-optional cursor
/// getters despite web's HoldingRow treating all three as nullable
/// (`exchange: string | null`, `instrument_type: string | null`,
/// `avg_cost: number | null`) -- a real runtime-crash risk for any
/// holding where those columns are actually null (e.g. a freshly-added
/// FD/crypto/other-scheme holding has no exchange). `current_value` was
/// already correctly typed `Int64?` but that's the one column this file
/// already read with the right optional accessor. Also added the three
/// columns the schema already had (PocketCareSchema.swift's `holdings`
/// table) but this repository's `Holding` struct was missing entirely:
/// `off_list` (functionally significant -- gates the "untracked" display,
/// see Domain/InvestmentsModel.swift's `holdingLabel`/`valuation`),
/// `source_account_id`, `planned_id`.
/// One `market_dividends` row (global, read-only market-sync table).
public struct DividendRow: Sendable {
    public let symbol: String
    public let exchange: String?
    public let exDate: String
    public let payDate: String?
    public let amount: Int64
    public let currency: String
}

/// One `market_quotes` row (global, read-only).
public struct QuoteRow: Sendable {
    public let symbol: String
    public let exchange: String?
    public let price: Int64
    public let currency: String
}

public struct Holding: Identifiable, Sendable {
    public let id: String
    public let userId: String
    public let accountId: String
    public let symbol: String
    public let exchange: String?
    public let quantity: Double
    public let avgCost: Int64?
    public let currency: String
    public let autoFetch: Bool
    public let instrumentType: String?
    public let offList: Bool
    public let name: String?
    public let assetClass: String?
    public let currentValue: Int64?
    public let annualRate: Double?
    public let maturityDate: String?
    public let sourceAccountId: String?
    public let plannedId: String?
    /// Amount-based SIP fields (migration 0061). `plannedId` alone is not
    /// enough to tell a live SIP from a stopped one -- it can still point at a
    /// recurring row that was already cancelled -- so web gates the SIP UI on
    /// `planned_id && sip_amount > 0` and so does mobile.
    public let sipAmount: Int64?
    public let sipDay: Int64?
}

/// Funding mode for a new holding -- matches web's AddHoldingInput.funding
/// union: an EXISTING holding raises the investment account's invested
/// pool via an `adjustment` entry (money was already in the market); a
/// NEW investment `transfer`s the cost from a chosen account into the
/// investment account (money moves, net worth preserved).
public enum HoldingFunding: Sendable {
    case existing
    case new(sourceAccountId: String)
}

/// A SIP the user is setting up alongside the holding -- web's
/// `AddHoldingInput.sip`, mirroring Android's `SipSetup`.
///
/// `firstDue` is the date the recurring engine posts the first instalment and
/// `startDate` is when the plan began; web sends the same value for both from
/// one date field, and they are kept separate here because the COLUMNS are
/// separate (`recurring_items.next_due` moves every time the engine posts,
/// `holdings.sip_start_date` never does).
///
/// `day` is the day-of-month, already clamped to 1-28 by Domain's
/// `clampSipDay` -- the repository does not re-clamp, so there is exactly one
/// place the rule lives.
public struct SipSetup: Sendable {
    public let amount: Int64
    public let frequency: String
    public let firstDue: String
    public let sourceAccountId: String
    public let startDate: String
    public let day: Int

    public init(amount: Int64, frequency: String, firstDue: String, sourceAccountId: String, startDate: String, day: Int) {
        self.amount = amount; self.frequency = frequency; self.firstDue = firstDue
        self.sourceAccountId = sourceAccountId; self.startDate = startDate; self.day = day
    }
}

public struct AddHoldingInput: Sendable {
    public let investmentAccountId: String
    public let assetClass: String
    public let symbol: String
    public let exchange: String?
    public let name: String
    public let quantity: Double
    public let avgCost: Int64?
    public let currency: String
    public let currentValue: Int64?
    public let annualRate: Double?
    public let maturityDate: String?
    public let offList: Bool
    public let autoFetch: Bool
    public let funding: HoldingFunding
    /// Present only when the user is starting a SIP. See `SipSetup`.
    public let sip: SipSetup?

    public init(investmentAccountId: String, assetClass: String, symbol: String, exchange: String?, name: String, quantity: Double, avgCost: Int64?, currency: String, currentValue: Int64?, annualRate: Double?, maturityDate: String?, offList: Bool, autoFetch: Bool, funding: HoldingFunding, sip: SipSetup? = nil) {
        self.investmentAccountId = investmentAccountId; self.assetClass = assetClass; self.symbol = symbol
        self.exchange = exchange; self.name = name; self.quantity = quantity; self.avgCost = avgCost
        self.currency = currency; self.currentValue = currentValue; self.annualRate = annualRate
        self.maturityDate = maturityDate; self.offList = offList; self.autoFetch = autoFetch
        self.funding = funding; self.sip = sip
    }
}

public actor InvestmentsRepository {
    private let db: PowerSyncDatabaseProtocol

    public init(db: PowerSyncDatabaseProtocol) {
        self.db = db
    }

    private func holdingMapper(cursor: SqlCursor) throws -> Holding {
        Holding(
            id: try cursor.getString(name: "id"),
            userId: try cursor.getString(name: "user_id"),
            accountId: try cursor.getString(name: "account_id"),
            symbol: try cursor.getString(name: "symbol"),
            exchange: try cursor.getStringOptional(name: "exchange"),
            quantity: try cursor.getDouble(name: "quantity"),
            avgCost: try cursor.getInt64Optional(name: "avg_cost"),
            currency: try cursor.getString(name: "currency"),
            autoFetch: (try cursor.getBooleanOptional(name: "auto_fetch")) ?? false,
            instrumentType: try cursor.getStringOptional(name: "instrument_type"),
            offList: (try cursor.getBooleanOptional(name: "off_list")) ?? false,
            name: try cursor.getStringOptional(name: "name"),
            assetClass: try cursor.getStringOptional(name: "asset_class"),
            currentValue: try cursor.getInt64Optional(name: "current_value"),
            annualRate: try cursor.getDoubleOptional(name: "annual_rate"),
            maturityDate: try cursor.getStringOptional(name: "maturity_date"),
            sourceAccountId: try cursor.getStringOptional(name: "source_account_id"),
            plannedId: try cursor.getStringOptional(name: "planned_id"),
            sipAmount: try cursor.getInt64Optional(name: "sip_amount"),
            sipDay: try cursor.getInt64Optional(name: "sip_day")
        )
    }

    public func watchHoldings(userId: String) throws -> AsyncThrowingStream<[Holding], Error> {
        try db.watch(
            sql: "SELECT * FROM holdings WHERE deleted_at IS NULL AND user_id = ? ORDER BY created_at DESC",
            parameters: [userId],
            mapper: holdingMapper
        )
    }

    /// Global, read-only market-sync tables -- no user_id/deleted_at filter,
    /// matches web's own unscoped queries exactly. Added 2026-08-06 for
    /// Insights' dividend_income/portfolio_projection cards (task #28), the
    /// first mobile reader of either table (mirrors Android's
    /// InvestmentsRepository.kt watchDividends()/watchQuotes() added the
    /// same session).
    public func watchDividends() throws -> AsyncThrowingStream<[DividendRow], Error> {
        try db.watch(
            sql: "SELECT symbol, exchange, ex_date, pay_date, amount, currency FROM market_dividends",
            parameters: []
        ) { cursor in
            DividendRow(
                symbol: try cursor.getString(name: "symbol"), exchange: try cursor.getStringOptional(name: "exchange"),
                exDate: try cursor.getString(name: "ex_date"), payDate: try cursor.getStringOptional(name: "pay_date"),
                amount: try cursor.getInt64(name: "amount"), currency: try cursor.getString(name: "currency")
            )
        }
    }

    /// The same rows as `watchDividends`, read once.
    ///
    /// The Investments view model rebuilds on the holdings stream and pulls
    /// balances and rates one-shot at that moment (see its `rebuild`); a
    /// second live stream would put it on two clocks, and the dividend table
    /// is refreshed by a once-a-day market sync, so there is nothing to watch
    /// between rebuilds. Mirrors LedgerRepository's `accountBalances()`.
    public func dividendsOnce() async throws -> [DividendRow] {
        try await db.getAll(
            sql: "SELECT symbol, exchange, ex_date, pay_date, amount, currency FROM market_dividends",
            parameters: [],
            mapper: { cursor in
                DividendRow(
                    symbol: try cursor.getString(name: "symbol"), exchange: try cursor.getStringOptional(name: "exchange"),
                    exDate: try cursor.getString(name: "ex_date"), payDate: try cursor.getStringOptional(name: "pay_date"),
                    amount: try cursor.getInt64(name: "amount"), currency: try cursor.getString(name: "currency")
                )
            }
        )
    }

    public func watchQuotes() throws -> AsyncThrowingStream<[QuoteRow], Error> {
        try db.watch(
            sql: "SELECT symbol, exchange, price, currency FROM market_quotes",
            parameters: []
        ) { cursor in
            QuoteRow(
                symbol: try cursor.getString(name: "symbol"), exchange: try cursor.getStringOptional(name: "exchange"),
                price: try cursor.getInt64(name: "price"), currency: try cursor.getString(name: "currency")
            )
        }
    }

    /// Adds a holding -- matches write.ts's addHolding() exactly: funds the
    /// invested pool first (transfer from a source account, or an adjustment
    /// on the investment account itself for an already-owned holding),
    /// creates the SIP's recurring transfer if there is one, then inserts the
    /// `holdings` row.
    ///
    /// The SIP half was missing until now, and its absence was not cosmetic:
    /// `planned_id` was hardcoded null and `sip_amount`/`sip_day` were never
    /// written, so a SIP could not be created on a phone AT ALL and the Stop
    /// SIP control could only ever appear for a holding created on web
    /// (PARITY_AUDIT 2026-08-28, item 9).
    @discardableResult
    public func addHolding(userId: String, input: AddHoldingInput) async throws -> String {
        let costTotal = Int64((Double(input.avgCost ?? 0) * input.quantity).rounded())
        let ts = nowIso()

        if costTotal > 0 {
            switch input.funding {
            case .new(let sourceAccountId):
                try await db.execute(
                    sql: """
                        INSERT INTO transactions
                         (id,user_id,account_id,type,amount,currency,to_account_id,to_amount,description,occurred_at,created_at,updated_at)
                        VALUES (?,?,?,?,?,?,?,?,?,?,?,?)
                        """,
                    parameters: [
                        newId(), userId, sourceAccountId, "transfer", costTotal, input.currency,
                        input.investmentAccountId, costTotal,
                        "Invested in \(input.name.isEmpty ? (input.symbol.isEmpty ? "investment" : input.symbol) : input.name)",
                        ts, ts, ts,
                    ]
                )
            case .existing:
                try await db.execute(
                    sql: """
                        INSERT INTO transactions
                         (id,user_id,account_id,type,amount,currency,description,occurred_at,created_at,updated_at)
                        VALUES (?,?,?,?,?,?,?,?,?,?)
                        """,
                    parameters: [
                        newId(), userId, input.investmentAccountId, "adjustment", costTotal, input.currency,
                        "Existing investment: \(input.name.isEmpty ? (input.symbol.isEmpty ? "holding" : input.symbol) : input.name)",
                        ts, ts, ts,
                    ]
                )
            }
        }

        // The SIP is a recurring `saving` transfer: source account -> this
        // investment account, auto-posting on its own schedule.
        //
        // This is the ONLY place a recurring saving item is created on either
        // platform, exactly as on web. Recurring savings are not browsable
        // under Recurring, so a SIP belongs to the holding that funds it and
        // is created and stopped there (see stopSipForHolding). It still posts
        // through the shared engine and still shows in the "Due now" strip.
        //
        // One row, not a template + rule pair: `recurring_items` carries both
        // the schedule and the transaction detail (migration 0064).
        // Written through `insertRow`, not the hand-rolled SQL the rest of this
        // function still uses: it is the documented convention (WriteHelpers),
        // it fills id/user_id/timestamps, and it routes through `withLoading`
        // so the sync indicator sees the write.
        var plannedId: String? = nil
        if let sip = input.sip, sip.amount > 0, !sip.sourceAccountId.isEmpty {
            plannedId = try await insertRow(
                db: db, table: "recurring_items", userId: userId,
                // The columns web sends as an explicit null (category_id, note,
                // labels, split_*, last_generated, source_*) are simply not
                // listed: an omitted column on an INSERT is NULL, and a
                // heterogeneous Swift dictionary literal cannot carry a bare
                // `nil` without a contextual type. Same rows result.
                //
                // Int64, not Int: this schema's integers bind as Int64
                // everywhere else in :data (see RecurringRepository.Input.row),
                // and a bare literal would infer Int.
                values: [
                    "direction": "saving",
                    "name": input.name.isEmpty ? (input.symbol.isEmpty ? "SIP" : input.symbol) : input.name,
                    "amount": sip.amount,
                    "currency": input.currency,
                    "frequency": sip.frequency,
                    // Web hardcodes 1 here too -- every interval the UI offers
                    // is "every 1 <frequency>".
                    "interval_count": Int64(1),
                    "next_due": sip.firstDue,
                    "account_id": sip.sourceAccountId,
                    "to_account_id": input.investmentAccountId,
                    "auto_post": Int64(1),
                    "active": Int64(1),
                    "description": "SIP",
                ]
            )
        }

        let id = newId()
        // A SIP's money movement IS the recurring transfer, so its source
        // account is the debit account, not a one-off funding account -- web
        // resolves it in the same order.
        let fundingSourceId: String? = { if case .new(let s) = input.funding { return s }; return nil }()
        let sourceAccountId: String? = input.sip?.sourceAccountId ?? fundingSourceId
        let instrumentType: String? = input.assetClass == "mf" ? "mf" : (input.assetClass == "stock" ? "stock" : nil)
        // Amount-based SIP fields (migration 0061). Written together with
        // `planned_id`: the UI reads a live SIP as
        // `planned_id != nil && sip_amount > 0`, and half of that pair is what
        // made every mobile-created holding look SIP-less.
        let sipAmount: Int64? = plannedId == nil ? nil : input.sip?.amount
        let sipStartDate: String? = plannedId == nil ? nil : input.sip?.startDate
        let sipDay: Int64? = plannedId == nil ? nil : input.sip.map { Int64($0.day) }
        try await db.execute(
            sql: """
                INSERT INTO holdings
                 (id,user_id,account_id,symbol,exchange,name,quantity,avg_cost,currency,asset_class,instrument_type,
                  current_value,annual_rate,maturity_date,source_account_id,planned_id,off_list,auto_fetch,
                  sip_amount,sip_start_date,sip_day,created_at,updated_at)
                VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
                """,
            parameters: [
                id, userId, input.investmentAccountId, input.symbol, input.exchange,
                input.name.isEmpty ? nil : input.name, input.quantity, input.avgCost, input.currency,
                input.assetClass, instrumentType, input.currentValue, input.annualRate, input.maturityDate,
                sourceAccountId, plannedId, input.offList ? 1 : 0, input.autoFetch ? 1 : 0,
                sipAmount, sipStartDate, sipDay, ts, ts,
            ]
        )
        return id
    }

    /// Matches web's EditHolding.save(): quantity/avg_cost always
    /// editable; current_value/annual_rate only meaningful for unpriced
    /// holdings but harmless to write through for priced ones (web does
    /// the same -- updateRow always sends all four).
    public func updateHolding(id: String, quantity: Double, avgCost: Int64?, currentValue: Int64?, annualRate: Double?) async throws {
        let ts = nowIso()
        try await db.execute(
            sql: "UPDATE holdings SET quantity = ?, avg_cost = ?, current_value = ?, annual_rate = ?, updated_at = ? WHERE id = ?",
            parameters: [quantity, avgCost, currentValue, annualRate, ts, id]
        )
    }

    /// Stop the SIP attached to a holding -- web's `stopSipForHolding()`.
    ///
    /// The recurring transfer is a standing debit on a real account, and
    /// recurring savings are not browsable under Recurring, so if this is
    /// skipped there is nowhere left for the user to go and cancel it: the
    /// money keeps leaving every month for an investment that no longer
    /// exists. No-op when the holding never had one.
    public func stopSipForHolding(plannedId: String?) async throws {
        guard let plannedId else { return }
        try await softDelete(db: db, table: "recurring_items", id: plannedId)
    }

    /// Clears the holding's own SIP fields, so `sipAmount > 0` stops reporting
    /// a live SIP once the recurring row is gone. Matches web's
    /// `updateRow("holdings", id, { sip_amount: null, sip_day: null })`.
    public func clearSipFields(id: String) async throws {
        try await db.execute(
            sql: "UPDATE holdings SET sip_amount = NULL, sip_day = NULL, updated_at = ? WHERE id = ?",
            parameters: [nowIso(), id]
        )
    }

    /// Matches web's remove(): soft-delete the holding row, and kill its SIP
    /// first -- the funding transaction is deliberately NOT reversed (same
    /// accepted asymmetry as Goals' allocation-delete-doesn't-cascade,
    /// documented in docs/mobile/screen-specs/goals.md), but a live standing
    /// debit is not an asymmetry, it is money still moving.
    public func deleteHolding(id: String, plannedId: String?) async throws {
        try await stopSipForHolding(plannedId: plannedId)
        try await softDelete(db: db, table: "holdings", id: id)
    }
}
